File: apps/spatial-video-control-example-android/hub-sample-provider/src/debug/java/io/github/mesmerprism/rustyquest/connection_hub_sample/ConnectionHubDebugSurfaceService.java
package io.github.mesmerprism.rustyquest.connection_hub_sample;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

/** DUMP-gated, debug-only off-head lifecycle for the sample Hub surface. */
public final class ConnectionHubDebugSurfaceService extends Service {
    public static final String ACTION_START =
            "io.github.mesmerprism.rustyquest.connection_hub_sample.action.START_CONNECTION_HUB_DEBUG_SURFACE";
    public static final String ACTION_STOP =
            "io.github.mesmerprism.rustyquest.connection_hub_sample.action.STOP_CONNECTION_HUB_DEBUG_SURFACE";
    private static final String TAG = "RQConnectionHubSample";
    private static final String CHANNEL_ID = "connection-hub-sample-debug";
    private static final int NOTIFICATION_ID = 2601;

    private ConnectionHubSampleProvider provider;

    @Override public void onCreate() {
        super.onCreate();
        provider = new ConnectionHubSampleProvider(this, (phase, toggled) ->
                Log.i(TAG, "phase=" + phase.replace(' ', '_') + " toggled=" + toggled));
    }

    @Override public int onStartCommand(Intent intent, int flags, int startId) {
        String action = intent == null ? "" : intent.getAction();
        if (ACTION_START.equals(action)) {
            startForeground(NOTIFICATION_ID, notification());
            provider.start();
            Log.i(TAG, "operator_action=start");
            return START_NOT_STICKY;
        }
        if (ACTION_STOP.equals(action)) {
            provider.stop();
            Log.i(TAG, "operator_action=stop");
            stopForeground(STOP_FOREGROUND_REMOVE);
            stopSelfResult(startId);
            return START_NOT_STICKY;
        }
        Log.w(TAG, "operator_action=rejected");
        stopSelfResult(startId);
        return START_NOT_STICKY;
    }

    @Override public void onDestroy() {
        provider.stop();
        super.onDestroy();
    }

    @Override public IBinder onBind(Intent intent) {
        return null;
    }

    private Notification notification() {
        NotificationManager manager = getSystemService(NotificationManager.class);
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "Connection Hub sample (debug)",
                NotificationManager.IMPORTANCE_LOW);
        manager.createNotificationChannel(channel);
        return new Notification.Builder(this, CHANNEL_ID)
                .setSmallIcon(android.R.drawable.stat_notify_sync_noanim)
                .setContentTitle("Connection Hub sample")
                .setContentText("Debug surface provider is active")
                .setOngoing(true)
                .build();
    }
}

File: apps/spatial-video-control-example-android/app/src/debug/java/io/github/mesmerprism/rustyquest/spatial_video_control/ConnectionHubDebugSurfaceService.kt
package io.github.mesmerprism.rustyquest.spatial_video_control

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log

/**
 * Debug-only foreground owner for off-head Connection Hub surface testing.
 * The DUMP permission is enforced by the debug manifest before this exported
 * service is entered. Only the two fixed actions below are accepted.
 */
class ConnectionHubDebugSurfaceService : Service() {
  private var surfaceClient: ConnectionHubSurfaceClient? = null

  override fun onBind(intent: Intent?): IBinder? = null

  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    return when (intent?.action) {
      START_ACTION -> {
        startAsForeground()
        if (surfaceClient == null) {
          surfaceClient =
              ConnectionHubSurfaceClient(this, DebugConnectionHubSurfaceTarget()).also {
                it.start()
              }
          marker("started")
        } else {
          marker("already_started")
        }
        START_NOT_STICKY
      }
      STOP_ACTION -> {
        closeSurface("stopped_by_operator")
        stopSelfResult(startId)
        START_NOT_STICKY
      }
      else -> {
        marker("rejected_unknown_action")
        stopSelfResult(startId)
        START_NOT_STICKY
      }
    }
  }

  override fun onDestroy() {
    closeSurface("service_destroyed")
    super.onDestroy()
  }

  private fun startAsForeground() {
    val notificationManager = getSystemService(NotificationManager::class.java)
    notificationManager.createNotificationChannel(
        NotificationChannel(
            NOTIFICATION_CHANNEL_ID,
            "Connection Hub debug surface",
            NotificationManager.IMPORTANCE_LOW,
        )
    )
    val notification =
        Notification.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(android.R.drawable.stat_notify_sync)
            .setContentTitle("Connection Hub debug surface")
            .setContentText("Off-head media controls are available for qualification")
            .setCategory(Notification.CATEGORY_SERVICE)
            .setOngoing(true)
            .build()
    startForeground(NOTIFICATION_ID, notification)
  }

  private fun closeSurface(reason: String) {
    val prior = surfaceClient
    surfaceClient = null
    prior?.close()
    stopForeground(STOP_FOREGROUND_REMOVE)
    marker(reason)
  }

  private fun marker(status: String) {
    Log.i(TAG, "channel=rusty-connection-hub-debug-surface status=$status")
  }

  companion object {
    val START_ACTION =
        "${BuildConfig.APPLICATION_ID}.action.START_CONNECTION_HUB_DEBUG_SURFACE"
    val STOP_ACTION =
        "${BuildConfig.APPLICATION_ID}.action.STOP_CONNECTION_HUB_DEBUG_SURFACE"

    private const val TAG = "RqHubDebugSurface"
    private const val NOTIFICATION_CHANNEL_ID = "connection-hub-debug-surface"
    private const val NOTIFICATION_ID = 0x4348
  }
}
